### ISSUE TITLE
App crash, on archived version

### ISSUE THREAD

**Issue by aabanaag** [Alamofire/Alamofire#3294]
* [x] I've read, understood, and done my best to follow the [*CONTRIBUTING guidelines](https://github.com/Alamofire/Alamofire/blob/master/CONTRIBUTING.md).

## What did you do?

I experienced a crash but I can't replicate it locally only when I build the app and install it.

## What did you expect to happen?

The app should not crash in the first place

## What happened instead?

ℹ Please replace this with of what happened instead.  

## Alamofire Environment

**Alamofire version: 5.2.2**
**Xcode version: 11.6**
**Swift version: 5**
**Platform(s) running Alamofire:**
**macOS version running Xcode: 10.15.5**

## Demo Project

ℹ Please link to or upload a project we can download that reproduces the issue.

This is the stack trace from firebase crashlytics
```
// This is how I created or initialized sessionManager
internal lazy var sessionManager: Session = {
        let configuration = URLSessionConfiguration.default
        let delegate = Session.default.delegate
        let manager = Session(configuration: configuration, delegate: delegate, startRequestsImmediately: true, cachedResponseHandler: nil, eventMonitors: [AlamofireLogger()])

        return manager
}()
```
```
Crashed: org.alamofire.session.rootQueue
0  libsystem_platform.dylib       0x18dfcc3d0 os_unfair_lock_lock$VARIANT$armv81 + 16
1  Alamofire                      0x104c39fbc Request.lastRequest.getter + 127 (__hidden#2399_:127)
2  Alamofire                      0x104c77f58 _hidden#6970_ (__hidden#7540_)
3  Alamofire                      0x104c2d9a0 _hidden#1656_ (__hidden#855_)
4  libdispatch.dylib              0x18df8a9a8 _dispatch_call_block_and_release
5  libdispatch.dylib              0x18df8b524 _dispatch_client_callout
6  libdispatch.dylib              0x18df68b3c _dispatch_lane_serial_drain$VARIANT$armv81
7  libdispatch.dylib              0x18df6954c _dispatch_lane_invoke$VARIANT$armv81
8  libdispatch.dylib              0x18df7284c _dispatch_workloop_worker_thread
9  libsystem_pthread.dylib        0x18dfdcb74 _pthread_wqthread + 272
10 libsystem_pthread.dylib        0x18dfdf740 start_wqthread + 8
```

**Comment by jshier** [comment]
If you're able to reproduce it, can you create a test project we can use to test?

**Comment by aabanaag** [comment]
I could but as of now I can't, I tried creating a new build the same issue

```
Crashed: org.alamofire.session.rootQueue.serializationQueue
0  libsystem_platform.dylib       0x1b6bfc3d0 os_unfair_lock_lock$VARIANT$armv81 + 16
1  Alamofire                      0x104a75fbc Request.lastRequest.getter + 127 (__hidden#2399_:127)
2  Alamofire                      0x104ab2f74 _hidden#6964_ + 4367380340 (__hidden#7540_:4367380340)
3  Alamofire                      0x104ab1b20 _hidden#2134_ + 218 (__hidden#7540_:218)
4  Alamofire                      0x104a8d0b4 _hidden#5162_ + 4367225012 (__hidden#855_:4367225012)
5  Alamofire                      0x104a9e39c _hidden#5298_ + 4367295388 (__hidden#855_:4367295388)
6  Alamofire                      0x104a9e378 _hidden#5297_ + 583 (__hidden#5723_:583)
7  Alamofire                      0x104a699a0 _hidden#1656_ + 4367079840 (__hidden#855_:4367079840)
8  libdispatch.dylib              0x1b6bba9a8 _dispatch_call_block_and_release + 24
9  libdispatch.dylib              0x1b6bbb524 _dispatch_client_callout + 16
10 libdispatch.dylib              0x1b6b98b3c _dispatch_lane_serial_drain$VARIANT$armv81 + 564
11 libdispatch.dylib              0x1b6b99580 _dispatch_lane_invoke$VARIANT$armv81 + 448
12 libdispatch.dylib              0x1b6b98a0c _dispatch_lane_serial_drain$VARIANT$armv81 + 260
13 libdispatch.dylib              0x1b6b9954c _dispatch_lane_invoke$VARIANT$armv81 + 396
14 libdispatch.dylib              0x1b6ba284c _dispatch_workloop_worker_thread + 580
15 libsystem_pthread.dylib        0x1b6c0cb74 _pthread_wqthread + 272
16 libsystem_pthread.dylib        0x1b6c0f740 start_wqthread + 8
```

**Comment by jshier** [comment]
Can you provide a more full crash report? I'd like to see the type of crash and platform.

**Comment by aabanaag** [comment]
The last comment I posted is the latest stack trace. I can't find a way to download the full crash report from crashlytics but I took a screenshot of it. 

![image](https://user-images.githubusercontent.com/2245822/92208813-89201700-eebe-11ea-9a3e-8ef2533d6f88.png)

**Comment by aabanaag** [comment]
Here is another crash report screenshot

![image](https://user-images.githubusercontent.com/2245822/92208891-ae148a00-eebe-11ea-83bb-a073a51c5483.png)

**Comment by aabanaag** [comment]
Hello? Any update? This is the stack trace from Crashlytics.

```
Crashed: com.apple.main-thread
0  libswiftCore.dylib             0x191416428 bool swift::RefCounts<swift::SideTableRefCountBits>::doDecrement<(swift::PerformDeinit)1>(unsigned int) + 8
1  Alamofire                      0x1029a77d0 Request.deinit + 29 (__hidden#5723_:29)
2  Alamofire                      0x1029ae1a8 _hidden#5050_ + 4340720040
3  libswiftCore.dylib             0x191415714 _swift_release_dealloc + 28
4  Alamofire                      0x102979c74 _hidden#2574_ + 4340505716 (__hidden#855_:4340505716)
5  libswiftCore.dylib             0x191415714 _swift_release_dealloc + 28
6  libsystem_blocks.dylib         0x1d2f41784 _Block_release + 188
7  libdispatch.dylib              0x18d39e280 _dispatch_client_callout + 16
8  libdispatch.dylib              0x18d3805b8 _dispatch_main_queue_callback_4CF$VARIANT$armv81 + 856
9  CoreFoundation                 0x18d6e4c30 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 12
10 CoreFoundation                 0x18d6df0e8 __CFRunLoopRun + 2480
11 CoreFoundation                 0x18d6de200 CFRunLoopRunSpecific + 572
12 GraphicsServices               0x1a37d9598 GSEventRunModal + 160
13 UIKitCore                      0x18ffa4004 -[UIApplication _run] + 1052
14 UIKitCore                      0x18ffa95d8 UIApplicationMain + 164
15 Mycare tst                     0x102676f8c main + 15 (__hidden#295_:15)
16 libdyld.dylib                  0x18d3bd598 start + 4
```

**Comment by aabanaag** [comment]
```
Crashed: com.apple.main-thread
0  libswiftCore.dylib             0x1b8371d04 _swift_release_dealloc + 16
1  Alamofire                      0x100905c74 _hidden#2574_ + 4298808436 (__hidden#855_:4298808436)
2  Alamofire                      0x100905c74 _hidden#2574_ + 4298808436 (__hidden#855_:4298808436)
3  libswiftCore.dylib             0x1b8371d10 _swift_release_dealloc + 28
4  libsystem_blocks.dylib         0x1aa816a4c _Block_release + 148
5  libdispatch.dylib              0x1aa8ee524 _dispatch_client_callout + 16
6  libdispatch.dylib              0x1aa8a05b4 _dispatch_main_queue_callback_4CF$VARIANT$mp + 904
7  CoreFoundation                 0x1aaba67fc __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 12
8  CoreFoundation                 0x1aaba16d0 __CFRunLoopRun + 1724
9  CoreFoundation                 0x1aaba0ce8 CFRunLoopRunSpecific + 424
10 GraphicsServices               0x1b4ceb38c GSEventRunModal + 160
11 UIKitCore                      0x1aeccf444 UIApplicationMain + 1932
12 Mycare tst                     0x100422f8c main + 15 (__hidden#295_:15)
13 libdyld.dylib                  0x1aaa288f0 start + 4
```

**Comment by hernangonzalez** [comment]
I may be getting the same crash on my `Authenticator`. 

This is my *bt*:
```
* thread #4, queue = 'com.journey.session.request', stop reason = EXC_BAD_INSTRUCTION (code=EXC_I386_INVOP, subcode=0x0)
    frame #0: 0x00007fff20360517 libsystem_platform.dylib`_os_unfair_lock_recursive_abort + 23
    frame #1: 0x00007fff2035b1a5 libsystem_platform.dylib`_os_unfair_lock_lock_slow + 258
    frame #2: 0x00000001087446ed Journey`UnfairLock.lock(self=(unfairLock = 0x00007f9314f05da0)) at Protected.swift:101:9
    frame #3: 0x0000000108744740 Journey`protocol witness for Lock.lock() in conformance UnfairLock at <compiler-generated>:0
    frame #4: 0x00000001087443d6 Journey`Lock.around<Self>(closure=0x0000000108745570 Journey`partial apply forwarder for closure #1 () -> A1 in Alamofire.Protected.write<A>((inout A) -> A1) -> A1 at <compiler-generated>, self=) at Protected.swift:39:9
    frame #5: 0x00000001087454af Journey`Protected.write<T>(closure=0x00000001086ddcf0 Journey`reabstraction thunk helper <A where A: Alamofire.Authenticator> from @callee_guaranteed (@inout Alamofire.AuthenticationInterceptor<A>.(MutableState in _3C9F9DFE4B92DED5A2689265F324D782)) -> () to @escaping @callee_guaranteed (@inout Alamofire.AuthenticationInterceptor<A>.(MutableState in _3C9F9DFE4B92DED5A2689265F324D782)) -> (@out ())partial apply forwarder with unmangled suffix ".13" at <compiler-generated>, self=0x0000600000cedec0) at Protected.swift:153:14
    frame #6: 0x00000001086d9323 Journey`closure #1 in AuthenticationInterceptor.refresh(result=success, self=0x000060000260ba20) at AuthenticationInterceptor.swift:342:32
    frame #7: 0x00000001084a5fb6 Journey`thunk for @escaping @callee_guaranteed (@in_guaranteed Result<BearerCredential, Error>) -> () at <compiler-generated>:0
  * frame #8: 0x00000001084a5982 Journey`ApiAuthenticator.refresh(credential=(accessToken = ""), session=0x00006000014daf00, completion=0x00000001084a6410 Journey`partial apply forwarder for reabstraction thunk helper from @escaping @callee_guaranteed (@in_guaranteed Swift.Result<JourneyKit.BearerCredential, Swift.Error>) -> () to @escaping @callee_guaranteed (@guaranteed Swift.Result<JourneyKit.BearerCredential, Swift.Error>) -> () at <compiler-generated>, self=0x000060000260b2d0) at ApiAuthenticator.swift:27:13
    frame #9: 0x00000001084a5f4e Journey`protocol witness for Authenticator.refresh(_:for:completion:) in conformance ApiAuthenticator at <compiler-generated>:0
    frame #10: 0x00000001086d904d Journey`AuthenticationInterceptor.refresh(credential=, session=0x00006000014daf00, mutableState=Alamofire.AuthenticationInterceptor<JourneyKit.ApiAuthenticator>.MutableState @ 0x0000600000ceded8, self=0x000060000260ba20) at AuthenticationInterceptor.swift:341:23
    frame #11: 0x00000001086d7e7c Journey`closure #1 in AuthenticationInterceptor.adapt(mutableState=Alamofire.AuthenticationInterceptor<JourneyKit.ApiAuthenticator>.MutableState @ 0x0000600000ceded8, urlRequest=Foundation.URLRequest @ 0x00007000003223b0, session=0x00006000014daf00, completion=0x0000000108800c80 Journey`partial apply forwarder for closure #4 (Swift.Result<Foundation.URLRequest, Swift.Error>) -> () in Alamofire.Session.performSetupOperations(for: Alamofire.Request, convertible: Alamofire.URLRequestConvertible) -> () at <compiler-generated>, self=0x000060000260ba20) at AuthenticationInterceptor.swift:270:17
    frame #12: 0x00000001086d808d Journey`partial apply for closure #1 in AuthenticationInterceptor.adapt(_:for:completion:) at <compiler-generated>:0
    frame #13: 0x0000000108745552 Journey`closure #1 in Protected.write<T>(closure=0x00000001086d8070 Journey`partial apply forwarder for closure #1 (inout Alamofire.AuthenticationInterceptor<A>.(MutableState in _3C9F9DFE4B92DED5A2689265F324D782)) -> Alamofire.AuthenticationInterceptor<A>.(AdaptResult in _3C9F9DFE4B92DED5A2689265F324D782) in Alamofire.AuthenticationInterceptor.adapt(_: Foundation.URLRequest, for: Alamofire.Session, completion: (Swift.Result<Foundation.URLRequest, Swift.Error>) -> ()) -> () at <compiler-generated>, self=0x0000600000cedec0) at Protected.swift:153:23
    frame #14: 0x0000000108745591 Journey`partial apply for closure #1 in Protected.write<A>(_:) at <compiler-generated>:0
    frame #15: 0x00000001087443e4 Journey`Lock.around<Self>(closure=0x0000000108745570 Journey`partial apply forwarder for closure #1 () -> A1 in Alamofire.Protected.write<A>((inout A) -> A1) -> A1 at <compiler-generated>, self=) at Protected.swift:40:16
    frame #16: 0x00000001087454af Journey`Protected.write<T>(closure=0x00000001086d8070 Journey`partial apply forwarder for closure #1 (inout Alamofire.AuthenticationInterceptor<A>.(MutableState in _3C9F9DFE4B92DED5A2689265F324D782)) -> Alamofire.AuthenticationInterceptor<A>.(AdaptResult in _3C9F9DFE4B92DED5A2689265F324D782) in Alamofire.AuthenticationInterceptor.adapt(_: Foundation.URLRequest, for: Alamofire.Session, completion: (Swift.Result<Foundation.URLRequest, Swift.Error>) -> ()) -> () at <compiler-generated>, self=0x0000600000cedec0) at Protected.swift:153:14
    frame #17: 0x00000001086d74e3 Journey`AuthenticationInterceptor.adapt(urlRequest=Foundation.URLRequest @ 0x00007000003226b0, session=0x00006000014daf00, completion=0x0000000108800c80 Journey`partial apply forwarder for closure #4 (Swift.Result<Foundation.URLRequest, Swift.Error>) -> () in Alamofire.Session.performSetupOperations(for: Alamofire.Request, convertible: Alamofire.URLRequestConvertible) -> () at <compiler-generated>, self=0x000060000260ba20) at AuthenticationInterceptor.swift:252:54
    frame #18: 0x00000001086daecb Journey`protocol witness for RequestAdapter.adapt(_:for:completion:) in conformance AuthenticationInterceptor<A> at <compiler-generated>:0
    frame #19: 0x00000001087ff711 Journey`Session.performSetupOperations(request=0x00006000011cc7e0, convertible=(payload_data_0 = 0x726573752f697061, payload_data_1 = 0xec000000656d2f73, payload_data_2 = 0x00000001088cdb38 type metadata for Alamofire.UploadRequest, instance_type = JourneyKit.UserMeRequest, witness_table_URLRequestConvertible = 0x00000001088a7f00 Journey`protocol witness table for JourneyKit.UserMeRequest : Alamofire.URLRequestConvertible in JourneyKit), self=0x00006000014daf00) at Session.swift:1050:17
    frame #20: 0x00000001087fdb9b Journey`Session.performDataRequest(request=0x00006000011cc7e0, self=0x00006000014daf00) at Session.swift:995:9
    frame #21: 0x00000001087fd7d6 Journey`closure #1 in closure #1 in Session.perform(request=0x00006000011cc7e0, self=0x00006000014daf00) at Session.swift:983:49
    frame #22: 0x00000001086da460 Journey`thunk for @escaping @callee_guaranteed () -> () at <compiler-generated>:0
    frame #23: 0x0000000109088ef8 libdispatch.dylib`_dispatch_call_block_and_release + 12
    frame #24: 0x000000010908a130 libdispatch.dylib`_dispatch_client_callout + 8
    frame #25: 0x000000010908cfd3 libdispatch.dylib`_dispatch_continuation_pop + 577
    frame #26: 0x000000010908c08b libdispatch.dylib`_dispatch_async_redirect_invoke + 755
    frame #27: 0x000000010909e279 libdispatch.dylib`_dispatch_root_queue_drain + 334
    frame #28: 0x000000010909ed03 libdispatch.dylib`_dispatch_worker_thread2 + 127
    frame #29: 0x0000000109133b28 libsystem_pthread.dylib`_pthread_wqthread + 244
    frame #30: 0x0000000109132b3f libsystem_pthread.dylib`start_wqthread + 15
```

it may be related to me creating my own queues? I followed the guidelines to use a serial one on **root**, yet **request** and **serialization** are concurrent.

**Comment by hernangonzalez** [comment]
Removed the custom queues from session and that's not it. Maybe this one is easier to read.
<img width="550" alt="Screen Shot 2020-10-04 at 08 31 33" src="https://user-images.githubusercontent.com/1162963/95014260-1cfb1500-061c-11eb-8369-c827bfe8c95d.png">

**Comment by hernangonzalez** [comment]
I could help my case by dispatching the completion here:

```
func refresh(_ credential: BearerCredential, for session: Session, completion: @escaping (Result<BearerCredential, Error>) -> Void)
```

The deadlock is produced by calling the completion immediately.

**Comment by aabanaag** [comment]
Sorry I'm not really sure what caused the issue. Can you enlighten me?

**Comment by hernangonzalez** [comment]
I would recommend you strip down your networking till you find something that might make sense. Specially any threading/queueing you may be using (since this is a deadlock). 
Good luck.

**Comment by aabanaag** [comment]
oh no. It sounds like a lot of work. 

From the screenshots you gave I hope this is the one causing it

```
internal lazy var sessionManager: Session = {
        let configuration = URLSessionConfiguration.af.default
        
        let delegate = Session.default.delegate
        let manager = Session(configuration: configuration,
                              delegate: delegate,
                              startRequestsImmediately: true,
                              cachedResponseHandler: nil,
                              eventMonitors: [AlamofireLogger()])

        return manager
    }()

    internal lazy var interceptor: Interceptor = {
        let oAuthAdaptor = OAuthAdaptor(baseUrl: self.configuration.baseUrl, persistenceService: self.persistenceService)
        let oAuthRetrier = OAuthRetrier(apiManager: self, baseUrl: self.configuration.baseUrl, persistenceService: self.persistenceService)
        let interceptor = Interceptor(adapter: oAuthAdaptor, retrier: oAuthRetrier)
        return interceptor
    }()

// Then request looks like this
let request = sessionManager.request(
            url,
            method: method,
            parameters: parameters,
            encoding: encoding,
            headers: HTTPHeaders(headers ?? [:]),
            interceptor: interceptor
        ).validate()
```

**Comment by jshier** [comment]
@hernangonzalez Your crash seems to be a double lock within the `AuthenticationInterceptor`. As you've discovered, we (unintentionally) expect the `refresh` method of your `Authenticator` to be asynchronous. I'll see if I can get a fix out soon. In the meantime, an asynchronous dispatch of the completion handler should workaround the issue.

@aabanaag You have multiple different crashes. The first, with `os_unfair_lock_lock$VARIANT$armv81` at the top of the call stack, appears to be some issue within `os_unfair_lock` itself. Your other two are different, one in the Swift runtime's `deinit` machinery, the other in `dealloc` itself. However, since the Alamofire portions of those crashes are hidden, it's hard to investigate. If you can get those properly resymbolicated that would help a lot.

**Comment by aabanaag** [comment]
@jshier I will do that today. Thank you!

**Comment by jshier** [comment]
@hernangonzalez You may also need to update your version of Alamofire, as your call stack is calling `performDataRequest` for what looks like an `UploadRequest`, which was a bug fixed in 5.0.5.

**Comment by jshier** [comment]
@hernangonzalez The issue should be fixed on the `bug/authenticator-recursive-locking` branch, please give it a try!

**Comment by jshier** [comment]
@aabanaag Have you been able to update to Xcode 12 / Swift 5.3? I wonder if that may fix your issues.

**Comment by aabanaag** [comment]
I tried publishing a build via testflight and I'm unable to replicate it but what I did was use appstore signing option, but I'll check if it fixed the issue.

**Comment by aabanaag** [comment]
Hello, I just got back to working on my app, and I still have the issue but only on ad-hoc builds.

**Comment by aabanaag** [comment]
I still have no clue on how to fix this. About the structure of my app, I have an SDK framework contains network calls and any network-related process for our backend which has the Alamofire dependency then I installed that framework onto my app. I think the reason why my crashes are hidden is due to the fact that it is from a framework that I installed onto the app. Any ideas on how to make sure the crashes from a framework gets unsymbolicated?
